src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,277 @@
+
+    A key is a org_id, string pair, either represented as a
+    tuple e.g (1, "a"), or a string "1:a".
+
+    Initial mapping is org_id's to sets of strings:
+        { 1: {"a", "b", "c"}, 2: {"e", "f"} }
+
+    Resulting ids will be stored in self.results, which is a
+    mapping of org_id to string -> int mapping:
+        {
+            1: {"a": 10, "b": 11, "c": 12},
+            2: {"e": 13},
+        }
+
+    For any keys that we didn't find ids for in the current
+    step (and therefore weren't added to the results) , use
+    `get_unmapped_keys` to return a new KeyCollection that
+    can be used for the next step in getting those ids.
+    """
+
+    def __init__(self, mapping: Mapping[int, Set[str]]):
+        self.mapping = mapping
+
+    def as_tuples(self) -> MutableSequence[Tuple[int, str]]:
+        """
+        Returns all the keys, each key represented as tuple -> (1, "a")
+        """
+        key_pairs: MutableSequence[Tuple[int, str]] = []
+        for org_id in self.mapping:
+            key_pairs.extend([(org_id, string) for string in self.mapping[org_id]])
+
+        return key_pairs
+
+    def as_strings(self) -> MutableSequence[str]:
+        """
+        Returns all the keys, each key represented as string -> "1:a"
+        """
+        keys: MutableSequence[str] = []
+        for org_id in self.mapping:
+            keys.extend([f"{org_id}:{string}" for string in self.mapping[org_id]])
+
+        return keys
+
+
+KR = TypeVar("KR", bound="KeyResult")
+
+
+@dataclass(frozen=True)
+class KeyResult:
+    org_id: int
+    string: str
+    id: int
+
+    @classmethod
+    def from_string(cls: Type[KR], key: str, id: int) -> KR:
+        org_id, string = key.split(":")
+        return cls(int(org_id), string, id)
+
+
+class KeyResults:
+    def __init__(self) -> None:
+        self.results: MutableMapping[int, MutableMapping[str, int]] = {}
+
+    def add_key_results(self, key_results: Sequence[KeyResult]) -> None:
+        for key_result in key_results:
+            try:
+                self.results[key_result.org_id][key_result.string] = key_result.id
+            except KeyError:
+                self.results[key_result.org_id] = {key_result.string: key_result.id}
+
+    def get_mapped_results(self) -> MutableMapping[int, MutableMapping[str, int]]:
+        return self.results
+
+    def get_unmapped_keys(self, keys: KeyCollection) -> KeyCollection:
+        """
+        Takes a KeyCollection and compares it to the results. Returns
+        a new KeyCollection for any keys that don't have corresponding
+        ids in results.
+        """
+        unmapped_org_strings: MutableMapping[int, Set[str]] = defaultdict(set)
+        for org_id, strings in keys.mapping.items():
+            for string in strings:
+                try:
+                    self.results[org_id][string]
+                except KeyError:
+                    unmapped_org_strings[org_id].add(string)
+
```

Comment:
Same, please check if the key is in the dictionary instead of relying on the (more expensive) exception.